### PR TITLE
Create app: Added Material Data Assistant.

### PR DESCRIPTION
**Pull request by AleziaKurdis**
This PR add a "**Material Data Assistant**" to the UI of the Material Entity.
You can dynamically change the different Material properties using a new UI.
This is not replacing the current JSON editor, but it add another way to compose material without have to code a JSON using the documentation to know how it is working.

image

image

**Note: This is tool have some limitation.** 
_(There are thing that Material allows that is not supported by this Assistant. There is still the JSON editor to do more advanced things)_
- The Assistant button will be **available only for a single material** json (for multiple "materials" the JSON editor need to be used). It **won't be available** for "**hifi_shader_simple**" model. 
- Some maps are not supported by the Assistant: lightMap, bumpMap, glossMap (you will need to use the JSON editor for these)

**Please, Test this carefully.**

_Also: See comment below about multiple material entities._

### CODE REVIEW

File: scripts/system/create/entityProperties/html/js/entityProperties.js [hunk 1]
```diff
@@ -4373,6 +4408,16 @@ function handleEntitySelectionUpdate(selections, isPropertiesToolUpdate) {
...
+        var materialDataText = JSON.stringify(materialJson);
+        if (getPropertyInputElement("materialURL").value === "materialData" && materialDataText.indexOf("hifi_shader_simple") === -1 &&
+                materialDataText.indexOf("glossMap") === -1 && materialDataText.indexOf("specularMap") === -1 && materialDataText.indexOf("bumpMap") === -1 && 
+                materialDataText.indexOf("lightMap") === -1  && materialDataText.indexOf("texCoordTransform0") === -1 && materialDataText.indexOf("texCoordTransform1") === -1 &&
+                (materialJson.materials === undefined || materialJson.materials.length <= 1 || typeof materialJson.materials === "object")) {
+            showMaterialAssistantButton();
+        } else {
+            hideMaterialAssistantButton();
+        }
+
```
Comment (ctrlaltdavid) [inline scripts/system/create/entityProperties/html/js/entityProperties.js:4420]: Would it be worth putting this into a function and reusing it in the same/similar code circa line 3530?

File: scripts/system/create/entityProperties/html/js/entityProperties.js [hunk 2]
```diff
@@ -4361,7 +4396,7 @@ function handleEntitySelectionUpdate(selections, isPropertiesToolUpdate) {
             showSaveMaterialDataButton();
             hideMaterialDataTextArea();
             hideNewJSONMaterialEditorButton();
-            hideMaterialDataSaved();
+            hideMaterialDataSaved();
```
Comment (ctrlaltdavid) [inline scripts/system/create/entityProperties/html/js/entityProperties.js:4399]:
```suggestion
            hideMaterialDataSaved();
```

File: scripts/system/html/css/materialAssistant.css [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    font-family: Raleway-Bold;
+    font-size: 18px; 
+    text-decoration: none; 
+    color:#F2F2F2;                
+}
+
+input[type=range].uiMaterialAssistant-slider {
+  margin: 10px 0;
+  width: 100%;
+  background-color: #2E2E2E;
```
Comment (ctrlaltdavid) [inline scripts/system/html/css/materialAssistant.css:93]: Inconsistent indent.

File: scripts/system/html/css/materialAssistant.css [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+div.uiMaterialAssistant-group{
+    background-color: #2E2E2E;
+    border: 2px solid #aaaaaa;
+    padding: 5px;
+    margin: 0px 0px 6px 0px;
+    font-family: Raleway-SemiBold;  
+    text-decoration: none; 
+    font-size: 13px; 
+    color:#ffffff;
```
Comment (ctrlaltdavid) [inline scripts/system/html/css/materialAssistant.css:63]:
```suggestion
    color: #ffffff;
```
And similar elsewhere.

File: scripts/system/create/entityProperties/html/js/materialAssistant.js [hunk 5]
```diff
@@ -0,0 +1,890 @@
...
+            if ($(maEmissiveColorPickerID).attr('active') === 'true') {
+                maMaterialData.emissive = [(rgb.r/255), (rgb.g/255), (rgb.b/255)];
+                maGenerateJsonAndSave();
+            }
+        }
+    });
+}
+
+function loadDataInMaUi(materialDataPropertyValue) {
+    var entityMaterialData,materialDefinition;
```
Comment (ctrlaltdavid) [inline scripts/system/create/entityProperties/html/js/materialAssistant.js:394]:
```suggestion
    var entityMaterialData, materialDefinition;
```

File: scripts/system/create/entityProperties/html/js/materialAssistant.js [hunk 6]
```diff
@@ -0,0 +1,890 @@
...
+
+    if (materialDefinition.occlusionMap !== undefined) {
+        maMaterialData.occlusionMap = materialDefinition.occlusionMap;
+    } else {
+        maMaterialData.occlusionMap = "";
+    }
+    maOcclusionMap.value = maMaterialData.occlusionMap;
+    
+    //CULL FACE MODE
+     if (materialDefinition.cullFaceMode !== undefined) {
```
Comment (ctrlaltdavid) [inline scripts/system/create/entityProperties/html/js/materialAssistant.js:694]:
```suggestion
    if (materialDefinition.cullFaceMode !== undefined) {
```

File: scripts/system/create/entityProperties/html/js/materialAssistant.js [hunk 7]
```diff
@@ -0,0 +1,890 @@
...
+    switch(maMaterialData.opacityMapMode) {
+        case "OPACITY_MAP_OPAQUE":
+            maOpacityMapModeDont.checked = true;
+            break;
+        case "OPACITY_MAP_MASK":
+            maOpacityMapModeMask.checked = true;
+            break;
+        case "OPACITY_MAP_BLEND":
+            maOpacityMapModeBlend.checked = true;
+    }
```
Comment (ctrlaltdavid) [inline scripts/system/create/entityProperties/html/js/materialAssistant.js:586]: Should a `default:` case be added to catch any unexpected `opactityMapMode` values?
Similarly in other `switch` statements?

File: scripts/system/create/entityProperties/html/js/materialAssistant.js [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+    maOpacity.value = maMaterialData.opacity;
+    maOpacitySlider.value = Math.floor(maMaterialData.opacity * 1000);
+
+    if (materialDefinition.opacityMapMode !== undefined) {
+        maMaterialData.opacityMapMode = materialDefinition.opacityMapMode;
+    } else {
+        maMaterialData.opacityMapMode = "OPACITY_MAP_OPAQUE";
+    }
+    switch(maMaterialData.opacityMapMode) {
```
Comment (ctrlaltdavid) [inline scripts/system/create/entityProperties/html/js/materialAssistant.js:577]:
```suggestion
    switch (maMaterialData.opacityMapMode) {
```
And similar elsewhere.

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: scripts/system/create/entityProperties/html/js/entityProperties.js [hunk 9]
```diff
@@ -3509,6 +3525,16 @@ function setMaterialDataFromEditor(noUpdate, entityIDsToUpdate) {
     } else {
         updateProperty('materialData', text, false);
     }
+    
+    var matJson = JSON.parse(text);
+    if ( getPropertyInputElement("materialURL").value === "materialData" && text.indexOf("hifi_shader_simple") === -1 &&
```
[the remark this answers is not in the document]
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/entityProperties.js]: Solved.

File: scripts/system/create/entityProperties/html/js/entityProperties.js (hunk 1 above)
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/entityProperties.js]: I was doing that when I saw your comments.
Solved. Availabilitty conditions has been centralized in materialAssistant.js (whit other functions related to the Material Assistant)

File: scripts/system/create/entityProperties/html/js/entityProperties.js (hunk 2 above)
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/entityProperties.js]: Fixed.

File: scripts/system/create/entityProperties/html/js/materialAssistant.js (hunk 5 above)
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/materialAssistant.js]: Fixed

File: scripts/system/html/css/materialAssistant.css (hunk 4 above)
Comment (AleziaKurdis) [reply scripts/system/html/css/materialAssistant.css]: Fixed.

File: scripts/system/create/entityProperties/html/js/materialAssistant.js (hunk 6 above)
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/materialAssistant.js]: Fixed

File: scripts/system/create/entityProperties/html/js/materialAssistant.js (hunk 7 above)
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/materialAssistant.js]: I added default condition to the 3 switch cases.
It raises an alert an error message. But such case should never happen unless the code or the UI get broken.

File: scripts/system/create/entityProperties/html/js/materialAssistant.js (hunk 8 above)
Comment (AleziaKurdis) [reply scripts/system/create/entityProperties/html/js/materialAssistant.js]: Fixed on the 3 switch case.

**Comment by AleziaKurdis** [comment]
Recent change also include a fix for the Multiple selection that wasn't working.
We can now select many material entities and set the material data for all the selected one at the same time. (But in the case of a multiple selection, the assistant will always start from default value, as it would be tricky to determine which one would be the right one to used as starting point, we can only expect that they wont' match anyway.) We already do something like that when we select the zones for renderWithZones for a multiple selection.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-Interface-PR1658-774bc6e-_0f050b91.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-774bc6e-.tar_0f050b91.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-774bc6e-_0f050b91.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-774bc6e-.tar_0f050b91.gz

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-774bc6e-_0f050b91.dmg

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-774bc6e-_0f050b91.exe

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by Colorfingers** [comment]
Very cool addition!  It works great... I'm not much of a builder but I can see how this will be useful.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-Interface-PR1658-b909cdc-_8340b019.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-b909cdc-_8340b019.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-b909cdc-.tar_8340b019.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-b909cdc-.tar_8340b019.Z

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-b909cdc-_8340b019.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-.tar_c49cb4fe.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-_c49cb4fe.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-.tar_c49cb4fe.Z

**build (macOS-10.15, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-Interface-PR1658-57ce9eb-_c49cb4fe.dmg

**build (windows-2019, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-_c49cb4fe.exe

**build (macOS-10.15, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-_c49cb4fe.dmg

**build (self-hosted_debian-11_aarch64, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-.tar_c49cb4fe.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-_c49cb4fe.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1658/Vircadia-PR1658-57ce9eb-.tar_c49cb4fe.gz

**Pull request merged**